pydrofoil/real.py

```diff
@@ -0,0 +1,264 @@
+    #             break
+    #     # return i, s[:i]+s[i+1:]
+    #     if lead_zero_count == len(s)-1:
+    #         return Real(rbigint.fromint(0), rbigint.fromint(1))
+    #     num = rbigint.fromstr((s[:i] + s[i+1:])[lead_zero_count:] if (s[0] != "+" and s[0] != "-") else s[0]+(s[1:i] + s[i+1:])[lead_zero_count:]) if i < len(s)-1 else rbigint.fromstr(s)
+    #     dif = len(s)-1-i
+    #     if dif < 19:
+    #         den = rbigint.fromint(10**dif)
+    #     else:
+    #         den = den_of_fromstr(dif)
+    #     return Real(num, den)
+    #     # num = ((s[:i] + s[i+1:])[lead_zero_count:] if (s[0] != "+" and s[0] != "-") else s[0]+(s[1:i] + s[i+1:])[lead_zero_count:]) if i < len(s)-1 else (s)
+    #     # dif = len(s)-1-i
+    #     # if dif < 19:
+    #     #     den = rbigint.fromint(10**dif)
+    #     # else:
+    #     #     den = den_of_fromstr(dif)
+    #     # # return Real(num, den)
+    #     # return lead_zero_count, num, den.str()
+
+
+
+    def add(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.add(num_new_2)
+        return Real(num_new, den_new)
+    
+    def sub(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.sub(num_new_2)
+        return Real(num_new, den_new)
+    
+    def mul(self, other):
+        num_new = self.num.mul(other.num)
+        den_new = self.den.mul(other.den)
+        return Real(num_new, den_new)
+    
+    def div(self, other):
+        if self.num.int_eq(0) and not other.num.int_eq(0):
+            num_new = rbigint.fromint(0)
+            den_new = rbigint.fromint(1)
+        elif other.num.int_eq(0):
+            assert False, "ZerodivideError: denominator cannot be 0"
+        else:
+            num_new = self.num.mul(other.den)
+            den_new = self.den.mul(other.num)
+        if den_new.int_lt(0):
+            num_new = num_new.neg()
+            den_new = den_new.neg()
+        return Real(num_new, den_new)
+
+    def pow(self, n):
+        if isinstance(n, int):
+            if n == 0:
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            elif n < MININT or n > MAXINT:
+                assert False, "exponent is out of range of INT"
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, rbigint):
+            if n.int_eq(0):
+                return Real(rbigint.fromint(1), rbigint.fromint(1))
+            else:
+                n = rbigint.fromint(n)
+                num_new = self.num.pow(n)
+                den_new = self.den.pow(n)
+                return Real(num_new, den_new)
+        elif isinstance(n, Real):
```

Comment:
this case isn't needed, pow has the signature `(real, int) -> real`. I also think it's fine to only support `n` being an `int` (because exponents that don't fit into a machine int will simply lead to `pow` not terminating), and call `rbigint.int_pow` instead of converting to an `rbigint` first.

Reply:
ok sure, I'm going to change it